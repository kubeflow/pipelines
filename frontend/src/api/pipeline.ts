import { Parameter } from './parameter';

export class Pipeline {
  public id: number;
  public created_at: string;
  public name: string;
  public description?: string;
  public package_id: number;
  public schedule: string;
  public enabled: boolean;
  public enabled_at: string;
  public parameters?: Parameter[];

  constructor() {
    this.created_at = new Date().toISOString();
    this.name = '';
    this.description = '';
    this.package_id = -1;
    this.schedule = '';
    this.enabled = false;
    this.enabled_at = new Date().toISOString();
    this.parameters = [];
  }
}
