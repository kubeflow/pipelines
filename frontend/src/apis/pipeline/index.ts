// tslint:disable
/**
 * backend/api/pipeline.proto
 * No description provided (generated by Swagger Codegen https://github.com/swagger-api/swagger-codegen)
 *
 * OpenAPI spec version: version not set
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

export * from './api';
export * from './configuration';
