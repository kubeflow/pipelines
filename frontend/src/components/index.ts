import { MyTopBar } from './top-bar';

// add custom elements here
const elements = {
  MyTopBar,
};

function toKebabCase(name: string) {
  return name.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
}

// register all components as kebab case
Object.keys(elements)
  .forEach(key => {
    customElements.define(toKebabCase(key), elements[key])
  });
