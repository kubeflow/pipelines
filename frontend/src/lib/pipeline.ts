import { Parameter } from './parameter';

export interface Pipeline {
  createAt?: string;
  id?: number;
  author: string;
  description: string;
  ends: number;
  name: string;
  packageId: number;
  parameters: Parameter[];
  recurring: boolean;
  recurringIntervalHours: number;
  starts: number;
}
