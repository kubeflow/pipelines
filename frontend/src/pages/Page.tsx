/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { ToolbarProps } from '../components/Toolbar';
import { BannerProps } from '../components/Banner';
import { SnackbarProps } from '@material-ui/core/Snackbar';
import { DialogProps } from '../components/Router';
import { errorToMessage } from '../lib/Utils';

export interface PageProps extends RouteComponentProps {
  toolbarProps: ToolbarProps;
  updateBanner: (bannerProps: BannerProps) => void;
  updateDialog: (dialogProps: DialogProps) => void;
  updateSnackbar: (snackbarProps: SnackbarProps) => void;
  updateToolbar: (toolbarProps: Partial<ToolbarProps>) => void;
}

export abstract class Page<P, S> extends React.Component<P & PageProps, S> {
  protected _isMounted = true;

  constructor(props: any) {
    super(props);
    this.props.updateToolbar(this.getInitialToolbarState());
  }

  public abstract render(): JSX.Element;

  public abstract getInitialToolbarState(): ToolbarProps;

  public abstract refresh(): Promise<void>;

  public componentWillUnmount(): void {
    this.clearBanner();
    this._isMounted = false;
  }

  public componentDidMount(): void {
    this.clearBanner();
  }

  public clearBanner(): void {
    this.props.updateBanner({});
  }

  public async showPageError(message: string, error?: Error, mode?: 'error' | 'warning'): Promise<void> {
    const errorMessage = await errorToMessage(error);
    this.props.updateBanner({
      additionalInfo: errorMessage ? errorMessage : undefined,
      message: message + (errorMessage ? ' Click Details for more information.' : ''),
      mode: mode || 'error',
      refresh: this.refresh.bind(this),
    });
  }

  public showErrorDialog(title: string, content: string): void {
    this.props.updateDialog({
      buttons: [{ text: 'Dismiss' }],
      content,
      title,
    });
  }

  protected setStateSafe(newState: Partial<S>, cb?: () => void): void {
    if (this._isMounted) {
      this.setState(newState as any, cb);
    }
  }
}
