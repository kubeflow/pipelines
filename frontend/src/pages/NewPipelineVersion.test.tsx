/*
 * Copyright 2019 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import NewPipelineVersion, { ImportMethod } from './NewPipelineVersion';
import TestUtils from '../TestUtils';
import { shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import { PageProps } from './Page';
import { Apis } from '../lib/Apis';
import { RoutePage, QUERY_PARAMS } from '../components/Router';
import { ApiResourceType } from '../apis/pipeline';

class TestNewPipelineVersion extends NewPipelineVersion {
  public _pipelineSelectorClosed = super._pipelineSelectorClosed;
  public _onDropForTest = super._onDropForTest;
}

describe('NewPipelineVersion', () => {
  let tree: ReactWrapper | ShallowWrapper;

  const historyPushSpy = jest.fn();
  const historyReplaceSpy = jest.fn();
  const updateBannerSpy = jest.fn();
  const updateDialogSpy = jest.fn();
  const updateSnackbarSpy = jest.fn();
  const updateToolbarSpy = jest.fn();

  let getPipelineSpy: jest.SpyInstance<{}>;
  let createPipelineSpy: jest.SpyInstance<{}>;
  let createPipelineVersionSpy: jest.SpyInstance<{}>;
  let uploadPipelineSpy: jest.SpyInstance<{}>;

  let MOCK_PIPELINE = {
    id: 'original-run-pipeline-id',
    name: 'original mock pipeline name',
    default_version: {
      id: 'original-run-pipeline-version-id',
      name: 'original mock pipeline version name',
      resource_references: [
        {
          key: {
            id: 'original-run-pipeline-id',
            type: ApiResourceType.PIPELINE,
          },
          relationship: 1,
        },
      ],
    },
  };

  let MOCK_PIPELINE_VERSION = {
    id: 'original-run-pipeline-version-id',
    name: 'original mock pipeline version name',
    description: 'original mock pipeline version description',
    resource_references: [
      {
        key: {
          id: 'original-run-pipeline-id',
          type: ApiResourceType.PIPELINE,
        },
        relationship: 1,
      },
    ],
  };

  function generateProps(search?: string): PageProps {
    return {
      history: { push: historyPushSpy, replace: historyReplaceSpy } as any,
      location: {
        pathname: RoutePage.NEW_PIPELINE_VERSION,
        search: search,
      } as any,
      match: '' as any,
      toolbarProps: TestNewPipelineVersion.prototype.getInitialToolbarState(),
      updateBanner: updateBannerSpy,
      updateDialog: updateDialogSpy,
      updateSnackbar: updateSnackbarSpy,
      updateToolbar: updateToolbarSpy,
    };
  }

  beforeEach(() => {
    jest.clearAllMocks();
    getPipelineSpy = jest
      .spyOn(Apis.pipelineServiceApi, 'getPipeline')
      .mockImplementation(() => MOCK_PIPELINE);
    createPipelineVersionSpy = jest
      .spyOn(Apis.pipelineServiceApi, 'createPipelineVersion')
      .mockImplementation(() => MOCK_PIPELINE_VERSION);
    createPipelineSpy = jest
      .spyOn(Apis.pipelineServiceApi, 'createPipeline')
      .mockImplementation(() => MOCK_PIPELINE);
    uploadPipelineSpy = jest.spyOn(Apis, 'uploadPipeline').mockImplementation(() => MOCK_PIPELINE);
  });

  afterEach(async () => {
    // unmount() should be called before resetAllMocks() in case any part of the unmount life cycle
    // depends on mocks/spies
    if (tree) {
      await tree.unmount();
    }
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  // New pipeline version page has two functionalities: creating a pipeline and creating a version under an existing pipeline.
  // Our tests will be divided into 3 parts: switching between creating pipeline or creating version; test pipeline creation; test pipeline version creation.

  describe('switching between creating pipeline and creating pipeline version', () => {
    it('creates pipeline is default when landing from pipeline list page', () => {
      tree = shallow(<TestNewPipelineVersion {...generateProps()} />);

      // When landing from pipeline list page, the default is to create pipeline
      expect(tree.state('newPipeline')).toBe(true);

      // Switch to create pipeline version
      tree.find('#createPipelineVersionUnderExistingPipelineBtn').simulate('change');
      expect(tree.state('newPipeline')).toBe(false);

      // Switch back
      tree.find('#createNewPipelineBtn').simulate('change');
      expect(tree.state('newPipeline')).toBe(true);
    });

    it('creates pipeline version is default when landing from pipeline details page', () => {
      tree = shallow(
        <TestNewPipelineVersion
          {...generateProps(`?${QUERY_PARAMS.pipelineId}=${MOCK_PIPELINE.id}`)}
        />,
      );

      // When landing from pipeline list page, the default is to create pipeline
      expect(tree.state('newPipeline')).toBe(false);

      // Switch to create pipeline version
      tree.find('#createNewPipelineBtn').simulate('change');
      expect(tree.state('newPipeline')).toBe(true);

      // Switch back
      tree.find('#createPipelineVersionUnderExistingPipelineBtn').simulate('change');
      expect(tree.state('newPipeline')).toBe(false);
    });
  });

  describe('creating version under an existing pipeline', () => {
    it('does not include any action buttons in the toolbar', async () => {
      tree = shallow(
        <TestNewPipelineVersion
          {...generateProps(`?${QUERY_PARAMS.pipelineId}=${MOCK_PIPELINE.id}`)}
        />,
      );
      await TestUtils.flushPromises();

      expect(updateToolbarSpy).toHaveBeenLastCalledWith({
        actions: {},
        breadcrumbs: [{ displayName: 'Pipeline Versions', href: '/pipeline_versions/new' }],
        pageTitle: 'Upload Pipeline or Pipeline Version',
      });
      expect(getPipelineSpy).toHaveBeenCalledTimes(1);
    });

    it('allows updating pipeline version name', async () => {
      tree = shallow(
        <TestNewPipelineVersion
          {...generateProps(`?${QUERY_PARAMS.pipelineId}=${MOCK_PIPELINE.id}`)}
        />,
      );
      await TestUtils.flushPromises();

      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineVersionName')({
        target: { value: 'version name' },
      });

      expect(tree.state()).toHaveProperty('pipelineVersionName', 'version name');
      expect(getPipelineSpy).toHaveBeenCalledTimes(1);
    });

    it('allows updating pipeline version description', async () => {
      tree = shallow(
        <TestNewPipelineVersion
          {...generateProps(`?${QUERY_PARAMS.pipelineId}=${MOCK_PIPELINE.id}`)}
        />,
      );
      await TestUtils.flushPromises();

      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineVersionDescription')({
        target: { value: 'some description' },
      });

      expect(tree.state()).toHaveProperty('pipelineVersionDescription', 'some description');
      expect(getPipelineSpy).toHaveBeenCalledTimes(1);
    });

    it('allows updating package url', async () => {
      tree = shallow(
        <TestNewPipelineVersion
          {...generateProps(`?${QUERY_PARAMS.pipelineId}=${MOCK_PIPELINE.id}`)}
        />,
      );
      await TestUtils.flushPromises();

      (tree.instance() as TestNewPipelineVersion).handleChange('packageUrl')({
        target: { value: 'https://dummy' },
      });

      expect(tree.state()).toHaveProperty('packageUrl', 'https://dummy');
      expect(getPipelineSpy).toHaveBeenCalledTimes(1);
    });

    it('allows updating code source', async () => {
      tree = shallow(
        <TestNewPipelineVersion
          {...generateProps(`?${QUERY_PARAMS.pipelineId}=${MOCK_PIPELINE.id}`)}
        />,
      );
      await TestUtils.flushPromises();

      (tree.instance() as TestNewPipelineVersion).handleChange('codeSourceUrl')({
        target: { value: 'https://dummy' },
      });

      expect(tree.state()).toHaveProperty('codeSourceUrl', 'https://dummy');
      expect(getPipelineSpy).toHaveBeenCalledTimes(1);
    });

    it("sends a request to create a version when 'Create' is clicked", async () => {
      tree = shallow(
        <TestNewPipelineVersion
          {...generateProps(`?${QUERY_PARAMS.pipelineId}=${MOCK_PIPELINE.id}`)}
        />,
      );
      await TestUtils.flushPromises();

      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineVersionName')({
        target: { value: 'test version name' },
      });
      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineVersionDescription')({
        target: { value: 'some description' },
      });
      (tree.instance() as TestNewPipelineVersion).handleChange('packageUrl')({
        target: { value: 'https://dummy_package_url' },
      });
      await TestUtils.flushPromises();

      tree.find('#createNewPipelineOrVersionBtn').simulate('click');
      // The APIs are called in a callback triggered by clicking 'Create', so we wait again
      await TestUtils.flushPromises();

      expect(createPipelineVersionSpy).toHaveBeenCalledTimes(1);
      expect(createPipelineVersionSpy).toHaveBeenLastCalledWith({
        code_source_url: '',
        name: 'test version name',
        description: 'some description',
        package_url: {
          pipeline_url: 'https://dummy_package_url',
        },
        resource_references: [
          {
            key: {
              id: MOCK_PIPELINE.id,
              type: ApiResourceType.PIPELINE,
            },
            relationship: 1,
          },
        ],
      });
    });

    // TODO(jingzhang36): test error dialog if creating pipeline version fails
  });

  describe('creating new pipeline', () => {
    it('renders the new pipeline page', async () => {
      tree = shallow(<TestNewPipelineVersion {...generateProps()} />);
      await TestUtils.flushPromises();
      expect(tree).toMatchSnapshot();
    });

    it('switches between import methods', () => {
      tree = shallow(<TestNewPipelineVersion {...generateProps()} />);

      // Import method is URL by default
      expect(tree.state('importMethod')).toBe(ImportMethod.URL);

      // Click to import by local
      tree.find('#localPackageBtn').simulate('change');
      expect(tree.state('importMethod')).toBe(ImportMethod.LOCAL);

      // Click back to URL
      tree.find('#remotePackageBtn').simulate('change');
      expect(tree.state('importMethod')).toBe(ImportMethod.URL);
    });

    it('creates pipeline from url', async () => {
      tree = shallow(<TestNewPipelineVersion {...generateProps()} />);

      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineName')({
        target: { value: 'test pipeline name' },
      });
      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineDescription')({
        target: { value: 'test pipeline description' },
      });
      (tree.instance() as TestNewPipelineVersion).handleChange('packageUrl')({
        target: { value: 'https://dummy_package_url' },
      });
      await TestUtils.flushPromises();

      tree.find('#createNewPipelineOrVersionBtn').simulate('click');
      // The APIs are called in a callback triggered by clicking 'Create', so we wait again
      await TestUtils.flushPromises();

      expect(tree.state()).toHaveProperty('newPipeline', true);
      expect(tree.state()).toHaveProperty('importMethod', ImportMethod.URL);
      expect(createPipelineSpy).toHaveBeenCalledTimes(1);
      expect(createPipelineSpy).toHaveBeenLastCalledWith({
        description: 'test pipeline description',
        name: 'test pipeline name',
        url: {
          pipeline_url: 'https://dummy_package_url',
        },
      });
    });

    it('creates pipeline from local file', async () => {
      tree = shallow(<NewPipelineVersion {...generateProps()} />);

      // Set local file, pipeline name, pipeline description and click create
      tree.find('#localPackageBtn').simulate('change');
      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineName')({
        target: { value: 'test pipeline name' },
      });
      (tree.instance() as TestNewPipelineVersion).handleChange('pipelineDescription')({
        target: { value: 'test pipeline description' },
      });
      const file = new File(['file contents'], 'file_name', { type: 'text/plain' });
      (tree.instance() as TestNewPipelineVersion)._onDropForTest([file]);
      tree.find('#createNewPipelineOrVersionBtn').simulate('click');

      tree.update();
      await TestUtils.flushPromises();

      expect(tree.state('importMethod')).toBe(ImportMethod.LOCAL);
      expect(uploadPipelineSpy).toHaveBeenLastCalledWith(
        'test pipeline name',
        'test pipeline description',
        file,
      );
      expect(createPipelineSpy).not.toHaveBeenCalled();
    });
  });
});
