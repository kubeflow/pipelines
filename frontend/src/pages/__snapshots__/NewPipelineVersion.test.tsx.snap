// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPipelineVersion creating new pipeline renders the new pipeline page 1`] = `
<div
  className="page"
>
  <div
    className="scrollContainer"
  >
    <div
      className="flex"
    >
      <WithStyles(WithFormControlContext(FormControlLabel))
        checked={true}
        control={
          <WithStyles(Radio)
            color="primary"
          />
        }
        id="createNewPipelineBtn"
        label="Create a new pipeline"
        onChange={[Function]}
      />
      <WithStyles(WithFormControlContext(FormControlLabel))
        checked={false}
        control={
          <WithStyles(Radio)
            color="primary"
          />
        }
        id="createPipelineVersionUnderExistingPipelineBtn"
        label="Create a new pipeline version under an existing pipeline"
        onChange={[Function]}
      />
    </div>
    <div
      className="explanation"
    >
      Upload pipeline with the specified package.
    </div>
    <Input
      autoFocus={true}
      id="newPipelineName"
      inputRef={
        Object {
          "current": null,
        }
      }
      label="Pipeline Name"
      onChange={[Function]}
      required={true}
      value=""
      variant="outlined"
    />
    <Input
      autoFocus={true}
      id="pipelineDescription"
      inputRef={
        Object {
          "current": null,
        }
      }
      label="Pipeline Description"
      onChange={[Function]}
      required={true}
      value=""
      variant="outlined"
    />
    <div
      className=""
    >
      URL must be publicly accessible.
    </div>
    <DocumentationCompilePipeline />
    <div
      className="flex"
    >
      <WithStyles(WithFormControlContext(FormControlLabel))
        checked={false}
        control={
          <WithStyles(Radio)
            color="primary"
          />
        }
        id="localPackageBtn"
        label="Upload a file"
        onChange={[Function]}
      />
      <n
        disableClick={true}
        disablePreview={false}
        disabled={true}
        getDataTransferItems={[Function]}
        id="dropZone"
        inputProps={
          Object {
            "tabIndex": -1,
          }
        }
        maxSize={Infinity}
        minSize={0}
        multiple={true}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDrop={[Function]}
        preventDropOnDocument={true}
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <Input
          InputProps={
            Object {
              "endAdornment": <WithStyles(InputAdornment)
                position="end"
              >
                <WithStyles(Button)
                  color="secondary"
                  disabled={true}
                  onClick={[Function]}
                  style={
                    Object {
                      "margin": 0,
                      "padding": "3px 5px",
                      "whiteSpace": "nowrap",
                    }
                  }
                >
                  Choose file
                </WithStyles(Button)>
              </WithStyles(InputAdornment)>,
              "readOnly": true,
              "style": Object {
                "maxWidth": 2000,
                "width": 455,
              },
            }
          }
          disabled={true}
          label="File"
          onChange={[Function]}
          required={true}
          value=""
          variant="outlined"
        />
      </n>
    </div>
    <div
      className="flex"
    >
      <WithStyles(WithFormControlContext(FormControlLabel))
        checked={true}
        control={
          <WithStyles(Radio)
            color="primary"
          />
        }
        id="remotePackageBtn"
        label="Import by url"
        onChange={[Function]}
      />
      <Input
        disabled={false}
        id="pipelinePackageUrl"
        label="Package Url"
        multiline={true}
        onChange={[Function]}
        style={
          Object {
            "maxWidth": 2000,
            "width": 465,
          }
        }
        value=""
        variant="outlined"
      />
    </div>
    <Input
      id="pipelineVersionCodeSource"
      label="Code Source (optional)"
      multiline={true}
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <div
      className="flex"
    >
      <BusyButton
        busy={false}
        className="buttonAction"
        disabled={true}
        id="createNewPipelineOrVersionBtn"
        onClick={[Function]}
        title="Create"
      />
      <WithStyles(Button)
        id="cancelNewPipelineOrVersionBtn"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(Button)>
      <div
        className="errorMessage"
      >
        Must specify either package url  or file in .yaml, .zip, or .tar.gz
      </div>
    </div>
  </div>
</div>
`;
